### PR TITLE
Add the ANY type and special null handling to the C API

### PR DESCRIPTION
**Pull request by taniabogatsch**
Specifically, this makes scalar UDFs more flexible.

### CODE REVIEW

**Comment by Giorgi** [comment]
> this makes scalar UDFs more flexible.

Can you clarify a bit what scenarios this PR enables? It's a little bit hard to understand for me by looking at the PR code.

Thanks

**Comment by taniabogatsch** [comment]
You can define scalar UDFs with `ANY` as the input type and various input parameters. That way, users can
- define a scalar UDF independent of input arguments (type and number).
A close example would be duckdb's `concat_ws` function.
```cpp
	ScalarFunction concat_ws = ScalarFunction("concat_ws", {LogicalType::VARCHAR, LogicalType::ANY},
	                                          LogicalType::VARCHAR, ConcatWSFunction, BindConcatWSFunction);
	concat_ws.varargs = LogicalType::ANY;
	concat_ws.null_handling = FunctionNullHandling::SPECIAL_HANDLING;
	set.AddFunction(concat_ws);
```

File: test/api/capi/capi_scalar_functions.cpp [hunk 1]
```diff
@@ -254,3 +254,83 @@ TEST_CASE("Test Scalar Functions - variadic number of input parameters", "[capi]
...
+		result_data[row_idx] = null_count;
+	}
+}
+
+static void CAPIRegisterANYFun(duckdb_connection connection, const char *name, duckdb_state expected_outcome) {
+	duckdb_state status;
+
+	// create a scalar function
+	auto function = duckdb_create_scalar_function();
+	duckdb_scalar_function_set_name(function, name);
```
Comment (Mytherin) [inline test/api/capi/capi_scalar_functions.cpp:290]: Could you add tests for these two functions, ensuring they fail when used with the `ANY` type (or INVALID)?

```cpp
duckdb_create_data_chunk
duckdb_bind_add_result_column
```

**Comment by taniabogatsch** [comment]
I've added significantly more tests for a lot of functions now.
What's left to do:
- [ ] A recurse check, if a `logical_type` contains a specified type. We need this for `duckdb_bind_add_result_column`, `duckdb_register_table_function`, and `duckdb_register_scalar_function`.

**Comment by taniabogatsch** [comment]
@Mytherin, I've added a `Contains` function to the `LogicalType`. Maybe you can have a look?
Otherwise, this now includes the requested tests, and fails with `INVALID` (and `ANY`) when registering scalar or table functions.

File: src/common/types.cpp [hunk 2]
```diff
@@ -1740,4 +1740,44 @@ bool LogicalType::operator==(const LogicalType &rhs) const {
 	return EqualTypeInfo(rhs);
 }
 
+bool LogicalType::Contains(const LogicalTypeId other) const {
+	switch (id()) {
+	case LogicalTypeId::STRUCT: {
```
Comment (Mytherin) [inline src/common/types.cpp:1745]: Can we use the `TypeVisitor` here?

**Review by Mytherin — COMMENTED** [review]
Thanks! LGTM - one minor comment

File: src/common/types.cpp (hunk 2 above)
Comment (taniabogatsch) [reply src/common/types.cpp:1745]: Oh, amazing, I was looking for something like this.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**